File: packages/python/src/agent_venv/events.py
"""Event log for environment lifecycles."""

from __future__ import annotations

import dataclasses
import time
from typing import Any, Literal, Protocol


EventKind = Literal[
    "env.created",
    "env.attached",
    "profile.materialized",
    "credentials.copied",
    "credentials.refreshed",
    "env.destroyed",
    "registry.read",
    "registry.written",
    "error",
]


@dataclasses.dataclass(slots=True)
class Event:
    ts_ms: int
    kind: EventKind
    data: dict[str, Any] = dataclasses.field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {"ts_ms": self.ts_ms, "kind": self.kind, **self.data}


class EventSink(Protocol):
    def __call__(self, event: Event) -> None: ...


class EventLog:
    def __init__(self, on_event: EventSink | None = None) -> None:
        self._origin = time.monotonic()
        self._events: list[Event] = []
        self._on_event = on_event

    def emit(self, kind: EventKind, /, **data: Any) -> Event:
        ts_ms = int((time.monotonic() - self._origin) * 1000)
        event = Event(ts_ms=ts_ms, kind=kind, data=data)
        self._events.append(event)
        if self._on_event is not None:
            try:
                self._on_event(event)
            except Exception:
                pass
        return event

    def all(self) -> list[Event]:
        return list(self._events)

File: packages/rust/src/events.rs
use serde_json::{Map, Value};
use std::sync::{Arc, Mutex};
use std::time::Instant;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EventKind {
    EnvCreated,
    EnvAttached,
    ProfileMaterialized,
    CredentialsCopied,
    CredentialsRefreshed,
    EnvDestroyed,
    RegistryRead,
    RegistryWritten,
    Error,
}

impl EventKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            EventKind::EnvCreated => "env.created",
            EventKind::EnvAttached => "env.attached",
            EventKind::ProfileMaterialized => "profile.materialized",
            EventKind::CredentialsCopied => "credentials.copied",
            EventKind::CredentialsRefreshed => "credentials.refreshed",
            EventKind::EnvDestroyed => "env.destroyed",
            EventKind::RegistryRead => "registry.read",
            EventKind::RegistryWritten => "registry.written",
            EventKind::Error => "error",
        }
    }
}

#[derive(Debug, Clone)]
pub struct Event {
    pub ts_ms: u64,
    pub kind: EventKind,
    pub data: Map<String, Value>,
}

impl Event {
    pub fn to_wire(&self) -> Map<String, Value> {
        let mut m = Map::new();
        m.insert("ts_ms".into(), Value::from(self.ts_ms));
        m.insert("kind".into(), Value::from(self.kind.as_str()));
        for (k, v) in &self.data {
            m.insert(k.clone(), v.clone());
        }
        m
    }
}

pub type EventSink = Arc<dyn Fn(&Event) + Send + Sync>;

pub(crate) struct EventLog {
    origin: Instant,
    events: Mutex<Vec<Event>>,
    sink: Option<EventSink>,
}

impl EventLog {
    pub fn new(sink: Option<EventSink>) -> Self {
        Self {
            origin: Instant::now(),
            events: Mutex::new(Vec::new()),
            sink,
        }
    }

    pub fn emit(&self, kind: EventKind, data: Map<String, Value>) {
        let ts_ms = self.origin.elapsed().as_millis() as u64;
        let event = Event { ts_ms, kind, data };
        if let Some(sink) = &self.sink {
            (sink)(&event);
        }
        if let Ok(mut guard) = self.events.lock() {
            guard.push(event);
        }
    }

    pub fn snapshot(&self) -> Vec<Event> {
        self.events.lock().map(|g| g.clone()).unwrap_or_default()
    }
}
